crates/generator/src/syscalls/redir_log.rs

```diff
@@ -0,0 +1,195 @@
+use gw_config::ContractLogConfig;
+use gw_types::{
+    bytes::{BufMut, BytesMut},
+    packed::RawL2Transaction,
+};
+use tokio::sync::mpsc;
+
+#[derive(Debug)]
+pub(crate) enum RedirLogMsg {
+    Session(RawL2Transaction),
+    Log(Vec<u8>),
+    Flush(i8), //exit code
+}
+pub(crate) struct RedirLogActor {
+    recv: mpsc::Receiver<RedirLogMsg>,
+    ctx: Context,
+}
+
+impl RedirLogActor {
+    pub(crate) fn new(recv: mpsc::Receiver<RedirLogMsg>, config: ContractLogConfig) -> Self {
+        let ctx = Context::init(config);
+        Self { recv, ctx }
+    }
+
+    fn handle_msg(&mut self, msg: RedirLogMsg) {
+        match msg {
+            RedirLogMsg::Session(tx) => self.ctx.setup(tx),
+            RedirLogMsg::Log(log) => self.ctx.append_log(&log),
+            RedirLogMsg::Flush(exit_code) => self.ctx.flush(exit_code),
+        }
+    }
+}
+
+// We can store the whole context of contract execution with tx, logs and exit code.
+struct Context {
+    tx: Option<RawL2Transaction>,
+    sink: BytesMut,
+    config: ContractLogConfig,
+}
+
+impl Context {
+    fn init(config: ContractLogConfig) -> Self {
+        Self {
+            tx: None,
+            sink: BytesMut::with_capacity(1024),
+            config,
+        }
+    }
+
+    fn setup(&mut self, tx: RawL2Transaction) {
+        self.tx = Some(tx);
+    }
+
+    fn append_log(&mut self, log: &[u8]) {
+        self.sink.put(log);
+        self.sink.put_u8(b'\n');
+    }
+
+    fn flush(&mut self, exit_code: i8) {
+        if let Ok(s) = std::str::from_utf8(&self.sink) {
+            if self.config == ContractLogConfig::Redirect
+                || (self.config == ContractLogConfig::RedirectError && exit_code != 0)
+            {
+                log::debug!("[contract debug]: {}", s);
+                log::debug!("contract exit code: {}", exit_code);
+            }
+            //send to senty if exit code != 0
+            //format:
+            //  tx_hash
+            //  [contrace logs]
+            //  ...
+            //  exit code: 3
+            if exit_code != 0 {
+                let mut entries: Vec<String> = Vec::with_capacity(3);
+                if let Some(tx) = &self.tx {
+                    let tx_hash = hex::encode(tx.as_reader().hash());
+                    entries.push(tx_hash);
+                }
+                entries.push(s.to_string());
+                entries.push(format!("exit code: {}", exit_code));
+                let msg = entries.join("\n");
```

Comment:
It should be possible to avoid allocation and concatenation? Push transaction hash and exit code to sink too and just `capture_msg(&s)` here.